File: app/src/main/java/moe/reimu/booru2pixiv/ShareHandlerActivity.java
package moe.reimu.booru2pixiv;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.Toast;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class ShareHandlerActivity extends Activity{
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this);

		final Uri data = getIntent().getData();

		OkHttpClient client = new OkHttpClient();

		Request request = new Request.Builder()
				.url(cleanupUri(data).buildUpon()
						.appendQueryParameter("login",
								sharedPref.getString(SettingsActivity.KEY_PREF_LOGIN_NAME, ""))
						.appendQueryParameter("api_key",
								sharedPref.getString(SettingsActivity.KEY_PREF_API_KEY, ""))
						.build().toString()
				).build();

		Log.d("B2P", request.url().toString());
		client.newCall(request).enqueue(new Callback(){

			@Override
			public void onFailure(Call call, IOException e) {
				showToast("API Request Failed");
				finish();
			}

			@Override
			public void onResponse(Call call, Response response) throws IOException {
				String jsonData = response.body().string();
				String pid = null;
				Log.d("B2P", jsonData);
				try	{
					JSONObject Jobject = new JSONObject(jsonData);
					if (!Jobject.getString("tag_string").contains("bad_id")) {
						pid = String.valueOf(Jobject.getInt("pixiv_id"));
					}
				} catch(Exception e) {
					Log.e("B2P", "JSON Parse failed");
				}

				if (pid == null){
					showToast("Pixiv ID not found / bad_id detected.");
					openLinkBrowser(data);
					finish();
					return;
				}


				Log.d("B2P", "PID=".concat(pid));
				showToast("Redirecting to Pixiv: ".concat(pid));
				openLinkApp(Uri.parse("http://www.pixiv.net/member_illust.php?mode=medium&illust_id=".concat(pid)));
				finish();
			}
		});
	}

	private Uri cleanupUri(Uri url){
		String ret = url.buildUpon().clearQuery().build().toString().concat(".json");
		return Uri.parse(ret);
	}

	private void openLinkBrowser(Uri url){
		Intent browserIntent = new Intent(Intent.ACTION_VIEW);
		browserIntent.setDataAndType(url, "text/html");
		browserIntent.addCategory(Intent.CATEGORY_BROWSABLE);
		startActivity(browserIntent);
	}

	private void openLinkApp(Uri url){
		Intent browserIntent = new Intent(Intent.ACTION_VIEW, url);
		startActivity(browserIntent);
	}

	private void showToast(final String text) {
		this.runOnUiThread(new Runnable() {
			public void run() {
				Toast.makeText(getApplicationContext(), text, Toast.LENGTH_SHORT).show();
			}
		});
	}

}

File: app/src/main/kotlin/moe/reimu/booru2pixiv/ShareHandlerActivity.kt
package moe.reimu.booru2pixiv

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.preference.PreferenceManager
import android.util.Log
import android.widget.Toast

import com.afollestad.materialdialogs.MaterialDialog

import moe.reimu.booru2pixiv.resolver.*


class ShareHandlerActivity : Activity() {
    companion object {
        private val TAG = "ShareHandlerActivity"

        var RESOLVE_RESULT = 1
        var RESOLVE_FAIL = 2
    }

    var progress: MaterialDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val sharedPref = PreferenceManager.getDefaultSharedPreferences(this)

        val origUri = intent.data
        val ctx = this
        showProgress()

        val handler = Handler(Handler.Callback { msg ->
            dismissProgress()

            if (msg.what == RESOLVE_FAIL) {
                showToast(getString(R.string.failed))
                openLinkBrowser(origUri)
                finish()
                return@Callback false
            }

            val result = msg.obj as Utils.Result

            val items = arrayOf(
                    "${ctx.getString(R.string.orig_uri)}\n($origUri)",
                    "${ctx.getString(result.sourceName)}\n(${result.uri})"
            )

            val dl = MaterialDialog.Builder(ctx)
                    .title(R.string.app_name)
                    .items(*items)
                    .itemsCallback { dialog, itemView, which, text ->
                        when(which) {
                            0 -> openLinkBrowser(origUri)
                            1 -> openLinkApp(result.uri)
                        }
                    }
                    .dismissListener { finish() }
                    .build()

            try {
                dl.show()
            } catch (e: Exception) {
                Log.d(TAG, "Activity closed?")
            }

            false
        })

        when (origUri.host) {
            Danbooru.HOST -> Danbooru().process(sharedPref, origUri, handler)
        }
    }


    private fun openLinkBrowser(url: Uri) {
        val browserIntent = Intent(Intent.ACTION_VIEW)
        browserIntent.setDataAndType(url, "text/html")
        browserIntent.addCategory(Intent.CATEGORY_BROWSABLE)
        startActivity(browserIntent)
    }

    private fun openLinkApp(url: Uri) {
        val browserIntent = Intent(Intent.ACTION_VIEW, url)
        startActivity(browserIntent)
    }

    private fun showToast(text: String) {
        this.runOnUiThread { Toast.makeText(applicationContext, text, Toast.LENGTH_SHORT).show() }
    }

    private fun showProgress() {
        progress = MaterialDialog.Builder(this)
                .title(R.string.app_name)
                .content(R.string.resolving)
                .progress(true, 0)
                .cancelListener { finish() }
                .show()
    }

    private fun dismissProgress() {
        this.runOnUiThread { progress?.dismiss() }
    }
}
